### ISSUE TITLE
dapr-placement seems to fail randomly

### ISSUE THREAD

**Issue by veselfil** [dapr/dapr#1732]
## In what area(s)?

/area placement

## What version of Dapr?

> 0.8.x

Running on AKS.

## Expected Behavior

Actors should get called when dapr-placement and the pods hosting the actor are running.

## Actual Behavior

Sometimes (seemingly quite randomly) the connection between dapr-placement and the daprd sidecars in the pods seems to fail. Always seems to happen for all pods in the entire cluster at the same time, suggesting the problem is probably in dapr-placement. Sometimes it fails in minutes, sometimes in days (right now we're on over 4 days, I don't think this has even happened before).

We have around 15 services, each using ~2 actors on average, all of them using the 0.8.0 version of the .NET DAPR SDK. However, we've had this issue ever since 0.5.1.

The error message we get from the SDK is "Error finding address for actor type {type}". We use GUIDs for our actors IDs, if that's relevant.

## Steps to Reproduce the Problem

Deploy DAPR to an AKS and wait. Unfortunately there doesn't seem to be anything triggering this.

**Comment by LMWF** [comment]
@veselfil is the error happening when you call ActorProxy.your_method_name()?


> We use GUIDs for our actors IDs, if that's relevant.

A guid is an ok id - this shouldn't be related.

**Comment by yaron2** [comment]
A few questions so we can try and reproduce:

1. Have you tried running in a different cluster to eliminate cluster network issues?
2. Which Kubernetes service/distro are you running?
3. Are there any special deployments running? For example, service mesh

**Comment by veselfil** [comment]
@LMWF yes, I'm pretty sure it is. It may also be when creating the actor proxy (`ActorProxy.Create<T>()`). We don't have a way of reproducing this reliably, so I'm not sure unfortunately. 

@yaron2 well, we're tried this in multiple AKS clusters of different versions. I don't think there are any network issues. Any idea how we should go about verifying that? 
We're not running any special deployments. No service meshes or anything of that sort.

Something interesting that happened - as I mentioned in the original post. We've moved over to a new cluster recently to which we started deploying our updates (we're talking probably close to 50 a day) and we've kept the old cluster, which wasn't getting new deployments anymore, for testing purposes until we made sure the new one was working as we wanted to. This led to dapr-placement working for a very long time (almost a week, much longer than usual). Could the problem maybe be some sort of a memory leak?

**Comment by veselfil** [comment]
Hi. We've set up monitoring of DAPR and the services in the cluster. We've tried to use a dashboard we've found online. I'm including a screenshot from Grafana:
![image](https://user-images.githubusercontent.com/6409595/88562524-1440fe00-d031-11ea-9134-2922b83f7500.png)

Unfortunately, nothing much to see there. As you can see, the error happened at 11:17 and I can't see anything strange happening with any of the metrics. Only thing is the number of goroutines reducing a little bit, though I think that's to be expected when a connection is closed. The actor runtime metrics (number of replicas etc...) haven't changed at all.

We're now settings up test services in our environment and collecting logs and metrics about the numbers of successful calls etc. to try to correlate the issue to *something*. Any tips on what we should focus on would be greatly appreciated.

Thanks.

**Comment by veselfil** [comment]
I should also add that we've since disabled mTLS. There was a theory that perhaps the issue was occuring when the certificates issued by sentry expired, but as we're still experiencing the problem, I'd say that rules this theory out.

**Comment by yaron2** [comment]
Do you see Pod restarts of the placement service?

**Comment by veselfil** [comment]
Yes, but only sometimes and not a large number of restarts.

**Comment by yaron2** [comment]
It would be extremely valuable to get the logs of the failed pod. You can get the logs with the --prrvious flag

**Comment by veselfil** [comment]
Right - I posted this on Gitter at first but I forgot to include them here, sorry for that. Here are the logs from the placement service:
```
{"instance":"dapr-placement-6795f8599-twqmq","level":"info","msg":"host removed: 10.244.3.218","scope":"dapr.placement","time":"2020-06-18T08:44:16.417661592Z","type":"log","ver":"0.8.0"}
{"instance":"dapr-placement-6795f8599-twqmq","level":"info","msg":"host removed: 10.244.3.217","scope":"dapr.placement","time":"2020-06-18T08:44:16.418349023Z","type":"log","ver":"0.8.0"}
{"instance":"dapr-placement-6795f8599-twqmq","level":"info","msg":"host removed: 10.244.3.194","scope":"dapr.placement","time":"2020-06-18T08:44:16.420011195Z","type":"log","ver":"0.8.0"}
{"instance":"dapr-placement-6795f8599-twqmq","level":"info","msg":"host removed: 10.244.3.194","scope":"dapr.placement","time":"2020-06-18T08:44:16.420131201Z","type":"log","ver":"0.8.0"}
{"instance":"dapr-placement-6795f8599-twqmq","level":"info","msg":"host removed: 10.244.3.218","scope":"dapr.placement","time":"2020-06-18T08:44:16.420203704Z","type":"log","ver":"0.8.0"}
{"instance":"dapr-placement-6795f8599-twqmq","level":"info","msg":"host removed: 10.244.3.217","scope":"dapr.placement","time":"2020-06-18T08:44:16.420232705Z","type":"log","ver":"0.8.0"}
{"instance":"dapr-placement-6795f8599-twqmq","level":"error","msg":"error updating host on lock operation: rpc error: code = Unavailable desc = transport is closing","scope":"dapr.placement","time":"2020-06-18T08:44:16.524647885Z","type":"log","ver":"0.8.0"}
{"instance":"dapr-placement-6795f8599-twqmq","level":"error","msg":"error updating host on update operation: rpc error: code = Unavailable desc = transport is closing","scope":"dapr.placement","time":"2020-06-18T08:44:16.549601779Z","type":"log","ver":"0.8.0"}
```

And here are the logs from one of the `daprd` containers: 

```
time="2020-06-18T08:11:52.791702843Z" level=warning msg="error establishing client to placement service: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 10.10.136.26:80: connect: connection refused\"" app_id=hcd-products-service instance=hcd-products-service-65fd5b59b6-lj66t scope=dapr.runtime.actor type=log ver=0.8.0
 time="2020-06-18T08:11:54.075795403Z" level=warning msg="error establishing client to placement service: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 10.10.136.26:80: connect: connection refused\"" app_id=hcd-products-service instance=hcd-products-service-65fd5b59b6-lj66t scope=dapr.runtime.actor type=log ver=0.8.0
 time="2020-06-18T08:11:56.088079052Z" level=warning msg="error establishing client to placement service: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 10.10.136.26:80: connect: connection refused\"" app_id=hcd-products-service instance=hcd-products-service-65fd5b59b6-lj66t scope=dapr.runtime.actor type=log ver=0.8.0
```

Important to note though, that we deleted the pod by hand (`kubectl delete pod`). The container itself seemed to be running fine, but the actor runtime wasn't working. Deleting the pod fixed it issue.

**Comment by yaron2** [comment]
Ok thanks.

So the pod didn't actually crash, it just aborted connections.

Those messages can be considered normal during a scaling operation on a deployment - do you manually and auto scale pods?

**Comment by veselfil** [comment]
Yes. I've never actually seen the pod in a `CrashLoopBackOff` state or anything similar. Always `Ready`, though sometimes it does have a restart or 2 on it.

We don't do any scaling per se, but we use rolling deployments and deploy very often.

**Comment by veselfil** [comment]
Another update - we're not running constant load against the services, calling 2 methods on an actor every 5 seconds. Unfortunately, the problem is still there, so it's probably not related to timeouts.

**Comment by veselfil** [comment]
Another update - pretty sure this is useful. It actually looks like there's not a hard cutoff for when it stops working. Have look at these metrics: 
![image](https://user-images.githubusercontent.com/6409595/88674201-d227c300-d0e9-11ea-930c-c08408529e02.png)

As I mentioned in the comment above, we're calling 2 methods, one monitored by the chart on the left, the other on the right. As you can see around 15:30, *some* of the calls started failing.

This conincides with a log message from dapr-placement: 
```
{\"instance\":\"dapr-placement-6694f95cf-dtwl5\",\"level\":\"error\",\"msg\":\"error updating host on unlock operation: rpc error: code = Unavailable desc = transport is closing\",\"scope\":\"dapr.placement\",\"time\":\"2020-07-28T13:29:19.644694734Z\",\"type\":\"log\",\"ver\":\"0.8.0\"}\n
```

Now for the new information: I noticed a restart on one of our pods. I had a look at the logs from the `previous` daprd container and found this on one of the replicas:
```
sync.(*Mutex).lockSlow(0xc0002f9e00)
        /opt/hostedtoolcache/go/1.14.3/x64/src/sync/mutex.go:138 +0xfc
sync.(*Mutex).Lock(...)
        /opt/hostedtoolcache/go/1.14.3/x64/src/sync/mutex.go:81
sync.(*RWMutex).Lock(0xc0002f9e00)
        /opt/hostedtoolcache/go/1.14.3/x64/src/sync/rwmutex.go:98 +0x97
github.com/dapr/dapr/pkg/actors.(*actorsRuntime).evaluateReminders(0xc000168b40)
        /home/runner/work/dapr/dapr/pkg/actors/actors.go:665 +0x57
created by github.com/dapr/dapr/pkg/actors.(*actorsRuntime).updatePlacements
        /home/runner/work/dapr/dapr/pkg/actors/actors.go:597 +0x4b6
```

On the other replica, the log was much, much longer, so I'm attaching it as a file. Hope this helps.
[log.txt](https://github.com/dapr/dapr/files/4988751/log.txt)

I think I was probably wrong in suggesting the problem is one placement's side, so sorry about that.

**Comment by yaron2** [comment]
This helps tremendously and really pinpoints the problem.

Thanks for that.

**Comment by yaron2** [comment]
Ok I can already see a place where we missed a lock that explains this.

**Comment by veselfil** [comment]
Awesome! Can't tell you how glad I am to hear that. Any idea when we might be able to get our hands on a fix?

**Comment by yaron2** [comment]
The PR is in: https://github.com/dapr/dapr/pull/1851

**Comment by yaron2** [comment]
The PR is now merged. you can use the image with the `:edge` tag to get a version with that fix.

It should fix the sidecar related issue, but we still have the placement pod restarting which also shouldn't happen.
Seeing the "host removed" and "connection closing" messages should be common in your case where you redeploy frequently, the the placement pod should not restart.

**Comment by veselfil** [comment]
Great, thanks! Is it possible to use the `edge` tag in the helm chart for installation? Is it fine to use the 0.9.0 version of the .NET SDK with the `edge` version?

I have a day off tomorrow, but I'll try to coordinate an update by the end of the week and I'll get back to you. It should be easy to see whether it helped with the monitoring we've got set up.

Thanks very much again.

**Comment by yaron2** [comment]
can't guarantee, but right now edge and 0.9.0 should be compatible.

**Comment by yaron2** [comment]
you can use the tag in the Helm chart, yes.

**Comment by veselfil** [comment]
Quick update - so far it looks to be more stable, haven't had a single failure on the updated version yet. We've been running it for about 30 hours, so not too long, but I'm cautiously optimistic.

**Comment by yaron2** [comment]
@veselfil great to read, waiting on further updates.

**Comment by yaron2** [comment]
@veselfil any updates?

**Comment by veselfil** [comment]
So far looks quite good. There are some very occasional failures (right now we're on over 400k successful invocations and 25 failed ones). Not sure what the cause of those is, I'll try to dig in the log once I have some time on my hands to try and find something that might help you. But the main issue sure looks to be fixed, so thanks for that.

**Comment by yaron2** [comment]
This is great to read, thanks.

occasional failures are to be expected whenever an actor fails over due to pod scaling in/out, but even that is an extremely low percentage! (0.006%).

If you can confirm you don't see the original cause anymore, please close the issue in your own free time.

**Comment by veselfil** [comment]
@yaron2 I have one more thing to ask - could you by any chance release the 0.9.0 version with the fix you've made? (as perhaps 0.9.1) We had to redeploy for unrelated reasons, and now we're obviously running on a different version of the DAPR runtime then when we deployed for the first time (using the `:edge` tag), which looks like it's causing some issues.

This would be immensely helpful to us. Thanks.

**Comment by veselfil** [comment]
@yaron2 any update please?

**Comment by yaron2** [comment]
> @yaron2 any update please?

I prepared the following image with 0.9.1: `yaron2/dapr:0.9.1`.

**Comment by veselfil** [comment]
Thanks for that. However, unfortunately, it doesn't seem to be working. I have no idea what has changed. A few calls go through sometimes after a restart and that's it. No errors in the logs, nothing. Interestingly, I've noticed there's `ver=edge` in the daprd logs. Could that maybe be a clue as to what's wrong?

**Comment by yaron2** [comment]
As this is a patch, there may have been other fixes were missed out.

However, can you give more details about "a few calls go through sometimes"? what does it means when a call doesn't go through? which status code are you getting back? what's the HTTP error status text/error message that gets returned?

**Comment by yaron2** [comment]
Also, make sure to apply this version you first:

1. Helm uninstall Dapr: `helm uninstall dapr -n dapr-system`
2. Install this version with `helm install dapr dapr/dapr --namespace dapr-system --set global.tag=0.9.1 --set `global.registry=yaron2`
3. Wait until the control plane is running and healthy with `dapr status -k`. 
4. Restart all the pods running Dapr. you can use the `kubectl rollout restart deploy/<deployment-name>` command.

**Comment by veselfil** [comment]
Hi @yaron2, sorry for not getting back to you, we've tried everything you said without success, so we have decided to wait for the 0.10 release. Hopefully the issue will be gone there.

**Comment by yaron2** [comment]
Hi @veselfil, have you tried the 0.10 release?

**Comment by KaiWalter** [comment]
@yaron2 On a cluster with **dapr 0.10.0** components and apps now being dormant for >1w I also experienced this error "Error finding address for actor type {type}" when restarting my test rig. Redeployment of apps did not help. After `kubectl rollout restart deployment.apps/dapr-placement -n dapr-system` and redeployment of the app the actors were working again.
Now I cannot generalize this problem as on the 2nd cluster in another region - which also had been dormant - everything got back to operation smoothly.
I'll keep an eye on this and will ask for re-opening this issue in case it manifests again.

**Comment by youngbupark** [comment]
> @yaron2 On a cluster with **dapr 0.10.0** components and apps now being dormant for >1w I also experienced this error "Error finding address for actor type {type}" when restarting my test rig. Redeployment of apps did not help. After `kubectl rollout restart deployment.apps/dapr-placement -n dapr-system` and redeployment of the app the actors were working again.
> Now I cannot generalize this problem as on the 2nd cluster in another region - which also had been dormant - everything got back to operation smoothly.
> I'll keep an eye on this and will ask for re-opening this issue in case it manifests again.

@KaiWalter I can reproduce it in my environment. I will investigate the issue.

**Comment by youngbupark** [comment]
Sharing the current progress, on local test, I can see unexpected EOF error from stream connection on placement server so that it keeps removing the Dapr sidecars from the membership list. trying keep-alive option doesn't work. I am still investigate the issue and will keep the progress posted here.

**Comment by youngbupark** [comment]
@KaiWalter I have the fix - #2071 

Uninstall Dapr first. Can you try to use my private build which includes the fix ?

```
helm install dapr dapr/dapr -n dapr-system --set-string global.tag=dev_20200914-linux-amd64 --set-string global.registry=docker.io/youngp --set-string dapr_placement.logLevel=debug
```

**Comment by KaiWalter** [comment]
@youngbupark did the reinstall but am not sure how to check the correct version. This is what is logged:
```
time="2020-09-18T10:44:37.689070264Z" level=info msg="starting Dapr Placement Service -- version edge -- commit v0.10.0-74-gfb3329d-dirty"
```

**Comment by youngbupark** [comment]
> @youngbupark did the reinstall but am not sure how to check the correct version. This is what is logged:
> 
> ```
> time="2020-09-18T10:44:37.689070264Z" level=info msg="starting Dapr Placement Service -- version edge -- commit v0.10.0-74-gfb3329d-dirty"
> ```

The version is correct. This is not the final fix. The final fix will be released to 0.11.0 Please let me know if you see any issues.

**Comment by KaiWalter** [comment]
@youngbupark have the fix now installed on our clusters - I suppose - and will monitor behavior

**Comment by veselfil** [comment]
Hi, it took us a long time to upgrade to DAPR 0.10, so apologies for my lack of activity. It seems we have now run into the same issue as @KaiWalter. We will also try upgrading to the custom build and see if it works.

**Comment by KaiWalter** [comment]
@youngbupark looks good on my end. After a few days and occasionally putting load on the actor scenarios, placement service / actors are still "alive".

**Comment by youngbupark** [comment]
> @youngbupark looks good on my end. After a few days and occasionally putting load on the actor scenarios, placement service / actors are still "alive".
> 
> 

Thanks for the report. This fix will go with the 0.11.0 release. 

JFYI - Apart from this fix, we are actively working on testing and improving the reliability of placement under very high load.(#2083) 

cc @msfussell @yaron2
